src/main/java/org/sonatype/nexus/blobstore/gcloud/internal/MultipartUploader.java

```diff
@@ -0,0 +1,256 @@
+ * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
+ *
+ * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
+ * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.blobstore.gcloud.internal;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.sonatype.nexus.blobstore.api.BlobStoreException;
+import org.sonatype.nexus.common.stateguard.StateGuardLifecycleSupport;
+
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.Storage.BlobTargetOption;
+import com.google.cloud.storage.Storage.ComposeRequest;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * Component that provides parallel multipart upload support for blob binary data (.bytes files).
+ */
+@Named
+public class MultipartUploader
+    extends StateGuardLifecycleSupport
+{
+
+  /**
+   * Use this property in 'nexus.properties' to control how large each multipart part is. Default is 5 MB.
+   * Smaller numbers increase the number of parallel workers used to upload a file. Match to your workload:
+   * if you are heavy in docker with large images, increase; if you are heavy in smaller components, decrease.
+   */
+  public static final String CHUNK_SIZE_PROPERTY = "nexus.gcs.multipartupload.chunksize";
+
+  /**
+   * This is a hard limit on the number of components to a compose request enforced by Google Cloud Storage API.
+   */
+  static final int COMPOSE_REQUEST_LIMIT = 32;
+
+  /**
+   * While an invocation of {@link #upload(Storage, String, String, InputStream)} is in-flight, the individual
+   * chunks of the file will have names like 'destination.chunkPartNumber", like
+   * 'content/vol-01/chap-01/UUID.bytes.chunk1', 'content/vol-01/chap-01/UUID.bytes.chunk2', etc.
+   */
+  private final String CHUNK_NAME_PART = ".chunk";
+
+  /**
+   * Used internally to count how many times we've hit the compose limit.
+   * Consider exposing this as a bean that can provide tuning feedback to deployers.
+   */
+  private final AtomicLong composeLimitHit = new AtomicLong(0);
```

Comment:
I'm thinking I will switch this to a Dropwizard `Counter` in the future and include it in the plugin's healthcheck.

Reply:
Makes sense, as long as `composeLimitHit` is just a metric and we're not using it to control behavior.